recap elixir deployment (#61)
dwyl/learn-devops: elixir-phoenix-deployement.md

# Deploy an Elixir or Phoenix application

> This markdown summarise the research done

Reviewer: ```suggestion
This markdown summarises the research done
```

on how to deploy a Phoenix application.

## PaaS vs VPS

> In the past we have mostly used Platform as a Service

Reviewer: ```suggestion
In the past we have mostly used Platform as a Service (PaaS) providers
```

> to deploy Elixir application. The main advantage of PaaS

Reviewer: ```suggestion
to deploy Elixir applications. The main advantage of PaaS
```

is it is quick to setup and have your application running
for your users.

> However on a long term using a PaaS can become costly.

Reviewer: ```suggestion
However in the longer term using a PaaS can become costly (per unit of compute e.g. RAM/CPU).
```

Virtual Private Server are in contrast normally cheaper to run

> but require more setup time and devops knowledge.

Reviewer: ```suggestion
but require more setup time and DevOps knowledge to maintain.
```

### PaaS

#### Heroku

We are used to deploy Elixir/Phoenix on Heroku
and we have already a ["how to"](https://github.com/dwyl/learn-phoenix-framework/blob/master/heroku-deployment.md) guide of this process.

However there are some limitations when using Heroku with Phoenix (see https://hexdocs.pm/phoenix/heroku.html):
![image](https://user-images.githubusercontent.com/6057298/83642784-2e1d1200-a5a7-11ea-8c97-9c7dd920469c.png)

pricing:
![image](https://user-images.githubusercontent.com/6057298/83641712-cc0fdd00-a5a5-11ea-8cbf-49981ae3747e.png)

The pricing above doesn't include the database which needs to be added to the total cost, see https://elements.heroku.com/addons/heroku-postgresql

### Gigalixir

Similar to Heroku [Gigalixir](https://www.gigalixir.com/) provides

> a platform which focus on deploying Elixir/Phoenix application.

Reviewer: ```suggestion
a platform which focuses _exclusively_ on deploying Elixir/Phoenix application.
Having an exclusive focus on deploying Elixir Apps gives Gigalixir several key advantages:
1. Zero-downtime blue-green continuous deployment.
see: https://martinfowler.com/bliki/BlueGreenDeployment.html
2. No limit to concurrent connections.
3. All clustering handled transparently. 
(so if you need to scale your app beyond 10k concurrent users, 
you don't have to pay exponentially more for bigger "dynos" 
the way you are forced to scale vertically on Heroku)
```

(so if you need to scale your app beyond 10k concurrent users, 
you don't have to pay exponentially more for bigger "dynos" 
the way you are forced to scale vertically on Heroku)

Deploying: https://elixircasts.io/deploying-with-gigalixir-%28revised%29

Pricing:

![image](https://user-images.githubusercontent.com/6057298/83643629-290c9280-a5a8-11ea-89da-648daec204f3.png)

Gigalixir vs Heroku:

![image](https://user-images.githubusercontent.com/6057298/83644108-d1225b80-a5a8-11ea-869a-c8c0e4a28012.png)

Like Heroku you need to add the cost for using the database:

![image](https://user-images.githubusercontent.com/6057298/83644296-09c23500-a5a9-11ea-9fdd-c50532195c95.png)

see also tiers pricing page: https://gigalixir.readthedocs.io/en/latest/tiers-pricing.html

### Render

[Render](https://render.com/) is another Paas similar to Heroku

pricing:

![image](https://user-images.githubusercontent.com/6057298/83645052-e9df4100-a5a9-11ea-8d3e-f1b18bb4cc02.png)

## Virtual Private Server

VPS allow us to manage ourself the deployement
setup. This allow us to customise the server and
the costs linked to it.

### Linode

#### Ubuntu

Linode provides and support Ubuntu:

> ![image](https://user-images.githubusercontent.com/6057298/83647811-3bd59600-a5ad-11ea-893c-b99e3df7f605.png)

Reviewer: ```suggestion
![linode-distribution-options-screen](https://user-images.githubusercontent.com/6057298/83647811-3bd59600-a5ad-11ea-893c-b99e3df7f605.png)
```

@SimonLab please get into the habit of _naming_ your images so that it's easier to review/maintain the docs. 😉 
(_and if you see me **not** naming my images, please point it out and/or fix it! thanks!_)

From there the idea is install Erlang and Elixir on the server
and then to run the application.

> - Install Erlang/Elixir with asdf:

Reviewer: @SimonLab as previously noted https://github.com/dwyl/phoenix-liveview-counter-tutorial/issues/17#issuecomment-625664043 please don't introduce a _completely_ new way of managing dependencies without giving _some_ context. This is the first mention of `asdf` in any file in the dwyl org. 

If you want to _propose_ using `asdf` https://github.com/asdf-vm/asdf as a version manager, and I have nothing against it in principal, it's as simple as opening an issue. see: https://github.com/dwyl/technology-stack/issues/18
Dropping a random tool into a doc without any context is not beginner friendly and sets an undesirable example to others who may think they can add _any_ dependency they like.

    - `git clone https://github.com/asdf-vm/asdf.git ~/.asdf --branch v0.7.8`
    - Edit `~/.bashrc file` and add `. $HOME/.asdf/asdf.sh` and run `source ~/.barhrc` to access the `asdf` command
    - Install required pacakges for Erlang `sudo apt install libssl-dev make automake autoconf libncurses5-dev gcc`

>     - Add erlang plugin to asdf: `asdf plugin-add erlang`

Reviewer: If we are using Linux as our Host OS then we 100% do _not_ need to install Erlang/Elixir on the VM. 
We can use Travis-CI which runs Ubuntu 18.04 ("Bionic") https://docs.travis-ci.com/user/reference/bionic
to build the package using [`distiliery`](https://github.com/bitwalker/distillery) and deploy it to the VM using [`edeliver`](https://github.com/edeliver/edeliver). 

> And as soon as Ubuntu `20.04 Focal` becomes available on Travis-CI  we can upgrade! Right now only `18.04` is supported as noted in: https://github.com/internetarchive/openlibrary/pull/3407#issuecomment-638637187
I fear that I may not have _fully_ explained "Continuous Integration + Continuous Deployment" in the issue ... #59  💭 
Perhaps I should have been more explicit that we would be following the "12 Factor App" best practices and separating the "Build Phase" from the "Run Phase" see: https://12factor.net/build-release-run
This strict separation ensures that there is no way to _modify_ the code on the production VM which _dramatically_ enhances security because it almost eliminates the risk of [arbitrary code execution](https://en.wikipedia.org/wiki/Arbitrary_code_execution) (_at least from the Erlang/Elixir perspective_).

The whole point is that our Continuous Integration server (Travis-CI) does the testing ***and*** build phase and then just _ships_ an **`executable binary`** of the application to the Linode VM. We _definitely_ don't want to be installing (_and thus having to **manage/maintain**_) **`asdf`** then `erlang` and `elixir` on the VM if we don't absolutely _need_ to.

We already demonstrated how to build and deploy a Phoenix App to a VM using [`distillery`](https://github.com/bitwalker/distillery) and [`edeliver`](https://github.com/edeliver/edeliver):
https://github.com/dwyl/learn-microsoft-azure#deploy-your-phoenix-web-app-using-edeliver
When we shipped [Healthlocker](https://github.com/healthlocker/healthlocker) to Azure in **2017**. 
See: https://github.com/healthlocker/healthlocker/blob/master/.deliver/config and: [`.travis.yml#L33-L35`](https://github.com/healthlocker/healthlocker/blob/becff6feafb3e10ebe061f7056bcc04bf4755c0c/.travis.yml#L33-L35)
This worked without any issues and we spent a _lot_ of time documenting all the scripts. e.g [`.deliver/version.sh`](https://github.com/healthlocker/healthlocker/blob/becff6feafb3e10ebe061f7056bcc04bf4755c0c/.deliver/version.sh)

The only reason I did not _specify_ using `edeliver` in the issue was because I wanted there to be a "fresh eyes" investigation into the _best_ way of doing Phoenix deployment to a VPS in 2020. _Installing_ the Erlang on the VPS and building the project on the production server is _definitely_ not the way we want to do it. The "_right_" way of doing this is to build the release (_binary_) on Travis-CI (_using someone else's compute time that they have generously provided for free!_) and uploading it to S3 so that we have a full history of all our release objects. Then deploy the release to the VPS which does _not_ have Erlang or _anything_ unnecessary installed on it. By _minimising_ the number of programs installed on the VM we minimise _maintenance_ and simplify our lives. The moment we introduce a new tool (e.g. `asdf`) we need to _understand_ the security+maintenance implications. If the `asdf` tool is installing any dependencies or running arbitrary shell scripts, who is _responsible_ for ensuring that it's not doing something undesirable (malicious) like opening a TCP port or leaking credentials? Where is `asdf` sourcing it's binaries for `erlang` and `elixir` from? Who has control over them? Is there a security audit? _Who_ is [Trevor Brown](https://github.com/Stratus3D) and _why_ should I trust them? He _seems_ like a "nice guy" https://www.linkedin.com/in/trevor-brown-11136938/ but how do we _know_ he is following security best practices cannot be ["spearphished"](https://en.wikipedia.org/wiki/Phishing#Spear_phishing), pwned and infect all the servers running `asdf`?

P.S. sorry to be a pain about this in the in the PR review, I should have made it clearer in the issue.
12 Factor App. Don't install build dependencies on the Production Server. 👍

    - Install Erlang: `asdf install erlang latest`
    - Install Elixir: `asdf install elixir latest`
    - Define which Elixir version to use `asdf global elxir <version>`

> - Instsall Nodejs using nvm

Reviewer: We _definitely_ don't want to install `node.js` or any of the node **build dependencies** on the Production Server. 
This is a _massive_ attack vector that would allow an attacker who successfully compromised one of the  _thousands_ of dependencies in the [`package-lock.json`](https://github.com/dwyl/hits/blob/master/assets/package-lock.json) file to create a `postinstall` script that leaks our AWS, DB or Encryption keys to a remote server. 
My (_security mindset_) reasoning for wanting to use `OpenBSD` on the Production server to _minimise_ the number of attack vectors would be _completely_ negated by installing `node.js` on the VPS. 

I _tolerate_ running Node.js on my `localhost` because I still maintain a bunch of NPM packages,
but I _definitely_ don't want to run Node.js on my Server unless I _have_ to (_e.g. because the Application is written in JS..._)

The following guide explain how to install FreeBSD on Linode:
https://www.linode.com/docs/tools-reference/custom-kernels-distros/install-freebsd-on-linode/

I've also tested the installation on one of my machine: https://github.com/SimonLab/FreeBSD-installation

The idea is then to use the FreeBSD package manager to install Elixir and Erlang:
- `pkg install erlang`
- `pkg install elixir`

### DigitalOcean

DigitalOcean provides a FreeBSD droplet:
![image](https://user-images.githubusercontent.com/6057298/83651322-472ac080-a5b1-11ea-8ce3-764cb9fe7927.png)

see https://www.digitalocean.com/products/linux-distribution/freebsd/

However OpenBSD can't be installed directly.
There are some way we could install it and investigate
if it can be used safely and without any blockers with DigitalOcean:
https://dev.to/nabbisen/custom-openbsd-droplet-on-digitalocean-4a9o

see also: https://www.digitalocean.com/community/tutorials/how-to-get-started-with-freebsd

## Current Conclusion

After reading and testing some Elixir/Phoenix/Linux/BSD installations
I can see that Linode (or similar) can be on a longer term a better tool
to manage the applications.

> The simple deployement used above with Ubuntu works well,

Reviewer: To be clear, the instructions given above are not a "Deployment" they are a "build and run on the server". That is _exactly_ what we _don't_ want to do and we _definitely_ don't want to advise other people to do it that way.

however I still have some research and testing to do especially
linked to continuous deployment without downtime.

> From reading the chapter 11 "Deploy Your Application to Production" of

Reviewer: Chapter 11 of the book focusses on using `mix release` which is why I asked if you had read it in: https://github.com/dwyl/learn-devops/issues/59#issuecomment-636433197
> "_Releases are produced with one of two tools—Mix or Distillery._"
![image](https://user-images.githubusercontent.com/194400/83729760-029a3600-a640-11ea-9e00-25bb650c47ef.png)

Real Time Phoenix, the solutions to deploy witout downtime are based on running
applications on Elixir clusters. I'd like to learn more about this aspect
but from a MVP perspective this point might take too much time to assimilate

> and I think a PaaS might be best to use at the moment.

Reviewer: PaaS is _definitely_ not how we are going to run the "Hits" application.
I'm not paying $200/month for a Heroku/AWS database just to let random GitHub users have a view counter on their repo! I want a commodity VPS solution.
The whole point of the issue https://github.com/dwyl/learn-devops/issues/59 was to figure out how to do deployment the ***the "right" (official) way***:
<img width="1231" alt="devops-issue-offiical-way" src="https://user-images.githubusercontent.com/194400/83730203-a257c400-a640-11ea-9b3c-42a956aa8e96.png">
That means using **`mix release`** as described in the official docs: https://hexdocs.pm/phoenix/releases.html

Perhaps _I_ should have been more "_strict_" and "course-corrected" when I saw your comment on using `asdf` to install dependencies on the VPS in https://github.com/dwyl/learn-devops/issues/59#issuecomment-635396537 ... I interpreted it as  you were "figuring things out" ... not that you were actually going to _manually_ install things on the VPS for a production deployment.

Reviewer: We ended up deploying Hits to Fly.io: https://github.com/dwyl/hits/pull/128 🚀
